xarray/tests/test_backends.py

```diff
@@ -6096,6 +6098,56 @@ def test_zarr_region_chunk_partial_offset(tmp_path):
         store, safe_chunks=False, region="auto"
     )
 
-    # This write is unsafe, and should raise an error, but does not.
-    # with pytest.raises(ValueError):
-    #     da.isel(x=slice(5, 25)).chunk(x=(10, 10)).to_zarr(store, region="auto")
+    with pytest.raises(ValueError):
+        da.isel(x=slice(5, 25)).chunk(x=(10, 10)).to_zarr(store, region="auto")
+
+
+@requires_zarr
+@requires_dask
+def test_zarr_safe_chunk(tmp_path):
+    # https://github.com/pydata/xarray/pull/8459#issuecomment-1819417545
+    store = tmp_path / "foo.zarr"
+    data = np.ones((20,))
+    da = xr.DataArray(data, dims=["x"], coords={"x": range(20)}, name="foo").chunk(x=5)
+
+    da.isel(x=slice(0, 7)).to_zarr(store, safe_chunks=True, mode="w")
+    with pytest.raises(ValueError):
+        # If the first chunk is smaller than the border size then raise an error
+        da.isel(x=slice(7, 11)).chunk(x=(2, 2)).to_zarr(
+            store, append_dim="x", safe_chunks=True
+        )
+
+    da.isel(x=slice(0, 7)).to_zarr(store, safe_chunks=True, mode="w")
+    # If the first chunk is of the size of the border size then it is valid
+    da.isel(x=slice(7, 11)).chunk(x=(3, 1)).to_zarr(
```

Comment:
Sorry if I'm being slow, but could I ask how this works? (It's a test so it must work, my understanding is lacking...)

IIUC, the existing array has chunks of (5,2). Then this adds a chunk of (3,1), making an array of (5,2,3,1). Is that correct? But don't the chunks need to be consistently sized?

Reply:
The array has 7 elements and chunks of size 5, at the moment of being written in Zarr it is going to generate one file/chunk with size 5 and the other with size 2, which means that the second chunk is incomplete, it needs 3 more elements to have size 5 (the missing elements to fill the chunk is what I call border size), and for that particular case the new array that is going to be appended have one chunk of size 3 and the other of size 1, which means that the first chunk is going to be inserted on the incomplete chunk and the second on a new chunk, for that reason it is safe to write without a synchronizer. 
I hope this helps to make me understand better.
![image](https://github.com/user-attachments/assets/b52d9d83-86eb-4e64-b956-1321db1b0633)

Reply:
Ah sorry, you're completely right. Thank you for the very clear explanation!